Add Concerns section in Refactoring portion of Getting Started guide (#39766)
rails/rails: guides/source/getting_started.md

The `@article` object is available to any partials rendered in the view because
we defined it as an instance variable.

### Using Concerns

>   A concern is any module that contains methods you would like to be able to share across multiple controllers or models. In other frameworks they are often known as mixins. They are a good way to keep your controllers and models small and keep your code DRY.

Reviewer: While this is accurate, concerns are also used more broadly for "extracting a slice of model that doesn’t seem part of its essence", to quote [DHH](https://signalvnoise.com/posts/3372-put-chubby-models-on-a-diet-with-concerns). Given that vision, we may not want to suggest in the guide that concerns are _only_ for DRYing up code.

Author: @dzunk I agree, but I'm not sure how to put that in a way that doesn't lead to more questions (in the article, DHH says " (what is and isn’t in the essence of a model is a fuzzy line and a longer discussion)" ) . 

perhaps something like "They are a good way to keep your controllers and models small and keep your code DRY, as well as extract any generic attributes?" 

Or perhaps I should remove this line entirely, and let the user infer when this might be useful? I'm not so sure about that. 🤔

Reviewer 2: This is a tough one to describe in a beginner friendly way :|

I'd describe concerns in Rails as an approach to make large models easier to understand and manage by extracting logic into well defined slices of responsibility. This comes with the benefit of reusability when multiple models share the same concerns.

Reviewer 2: This is a tough one to describe in a beginner friendly way :|

I'd describe concerns in Rails as an approach to make large models easier to understand and manage by extracting logic into well defined slices of responsibility. This comes with the benefit of reusability when multiple models share the same concerns.

Author: @fylooi Thanks for the help here! I've updated the guide with your suggestion, as well as amended a paragraph further down (which I will tag you on) . Would be great to hear your thoughts!

>  You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog` , two folders were created within `apps/` along with the rest:

Reviewer 3: ```suggestion
 You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog` , two folders were created within `app/` along with the rest:
```

 ```
 app/controllers/concerns
 app/models/concerns
 ```

> A given blog article might have various statuses - for instance, it may be `public` (visible to everyone), or `private` (only visible by the author, perhaps if an author wants to save a draft). It may also be `archived`, which means it is hidden from the author. A comment, similarly, may be `public`, `private`, or `archived`. For any given article, in the controller you might want to verify that an article is public before rendering it.

Reviewer 4: ```suggestion
A given blog article might have various statuses - for instance, it may be `public` (visible to everyone), or `private` (only visible to the author, perhaps if an author wants to save a draft). It may also be `archived`, which means it is hidden from the author. A comment, similarly, may be `public`, `private`, or `archived`. For any given article, in the controller you might want to verify that an article is public before rendering it.
```

  def archived?
    status == 'archived'
  end
end
```

Then, in our `index` action template (`app/views/articles/index.html.erb`) we would use the `archived?` method to avoid displaying any article that is archived:

```html+erb
<h1>Listing Articles</h1>
<%= link_to 'New article', new_article_path %>
<table>
  <tr>
    <th>Title</th>
    <th>Text</th>
    <th colspan="3"></th>
  </tr>

  <% @articles.each do |article| %>
    <% unless article.archived? %> 
      <tr>
        <td><%= article.title %></td>
        <td><%= article.text %></td>
        <td><%= link_to 'Show', article_path(article) %></td>
        <td><%= link_to 'Edit', edit_article_path(article) %></td>
        <td><%= link_to 'Destroy', article_path(article),
                method: :delete,
                data: { confirm: 'Are you sure?' } %></td>
      </tr>
    <% end %>
  <% end %>
</table>
```

However, if you look again at our models now, you can see that the logic is duplicated. If in future we increase the functionality of our blog - to include private messages, for instance -  we might find ourselves duplicating the logic yet again. This is where concerns come in handy.

> Let's call our new concern (module) `Statusable`, a module to use for any model that has a status. We can create a new file inside `app/models/concerns` called `statuslike.rb` , and store all of the status methods that were duplicated in the models.

Reviewer: `Statusable` and `Statuslike` are used inconsistently within these examples. I'd also suggest `Archivable` as a clearer name for this concern.

Author: Whoops! I intended to fully replace Statuslike with Statusable. Archivable is a pretty good name though! I'll probably go with that, though it might hide the fact that this also controls general visibility. Maybe `Viewable` , or is that too generic?

Reviewer: `Viewable` does sound a tad generic to me. We could also reduce the scope of the example to only focus on archiving, after all it's more about showing how to use concerns than the specific functionality they introduce.

Reviewer: In practice, I might add two concerns `Draftable` and `Archivable` to describe the public/private and published/archived functionality.

Author: @dzunk  Hmm, I can see where you're coming from, but I think that adding two concerns in this guide might be a bit confusing for a new reader. I think I might go along with one of the name examples in this [DHH](https://signalvnoise.com/posts/3372-put-chubby-models-on-a-diet-with-concerns) article - the `Visible` concern they mention. It's still somewhat generic but visible as a word most accurately reflects the intention behind the status (while actually being a proper word as well, unlike `Viewable`. ) What do you think?

`app/models/concerns/statuslike.rb`

```ruby
module Statuslike
  def archived?
    status == 'archived'
  end
end
```

> To include the duplicated validations, we will want to extend `ActiveSupport::Concern` . This module allows us to write validations and other Active Model methods in the module as if we were writing them within the original models (or controllers, if writing a concern for controllers). You can read more about this in the [API Guide](https://api.rubyonrails.org/classes/ActiveSupport/Concern.html):

Author: Updated! Thanks

```ruby
module Statuslike
  extends ActiveSupport::Concern

>   included do

Reviewer: It might be nice to include an example of `class_methods do...end` as well, as the second piece of the concern magic.

    VALID_STATUSES = ['public', 'private', 'archived']

    validates :status, in: VALID_STATUSES
  end

  def archived?
    status == 'archived'
  end
end
```

Now, we can remove the duplicated logic from each model and instead include our new `Statusable` module:

In `app/models/article.rb`:
```ruby
class Article < ApplicationRecord

>   includes Statusable

Reviewer 4: ```suggestion
  include Statusable
```

  has_many :comments

  validates :title, presence: true,
                    length: { minimum: 5 }

end
```

and in `app/models/comment.rb`:

```ruby
class Comment < ApplicationRecord

>   includes Statusable

Reviewer 4: ```suggestion
  include Statusable
```

  belongs_to :article
end
```

> Our app contains `app/models/concerns` and `app/controllers/concerns` in its `autoload_paths` so you don't need to `require` the concern. If your app is having trouble finding your concern module, you may need to check your `autoload_paths`. You can learn about how to do so [here](http://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoload-paths).

Author: Thanks for the feedback, I have removed it. I think I ran into an issue with this years ago as a beginner to Rails, so thought I'd add it in, but it's possible there was some issue with the default settings in Rails at the time.

The `@article` object is available to any partials rendered in the view because
we defined it as an instance variable.

### Using Concerns

> A concern is any module that contains methods representing a well-defined slice of the functionality that a model or controller is responsible for. In other frameworks they are often known as mixins. Concerns are a way to make large controllers or models easier to understand and manage. This also has the advantage of reusability when multiple models (or controllers) share the same concerns.

Reviewer 2: While accurate, I'd hesitate to describe a concern as a module.

I think concerns in Rails are primarily a design approach / style / pattern to break down models into separate (and reusable) slices of responsibility. Usage of modules is an implementation detail. 

Here's a [different approach](http://andrzejonsoftware.blogspot.com/2011/02/dci-and-rails.html) which is also implemented with modules containing well defined slices of functionality

```
class Article
end

module Notifier
  def send_notifications
    ...
  end
end

class ArticlesController
  def create
    @article.extends(Notifier)

    if @article.save
       @article.send_notifications
    end
  end
end

Reviewer 5: Perhaps reusing "a way to" makes it clearer that it's a pattern instead of tying it to modules?

```suggestion
A concern is a way to represent a well-defined slice of the functionality that a model or controller is responsible for. In other frameworks they are often known as mixins. They help make large controllers or models easier to understand and manage. This also has the advantage of reusability when multiple models (or controllers) share the same concerns.
```

Reviewer 2: > Perhaps reusing "a way to" makes it clearer that it's a pattern instead of tying it to modules?

I like the flow, think the mixin line is more of a reference to modules though.

Author: @fylooi @sunny I just made a new revision - Swapped around the order of the sentences. I'm not sure if the mixin line is still worth including or not as a definition of modules might not be needed in a Rails guide, but I've kept it so far because it seems like a useful comparison to make to quickly build up understanding for those that are new to Rails.

Reviewer 5: Thank you! 👏 

> I'm not sure if the mixin line is still worth including

I'd be inclined to remove it. To me, mixins are how we call "includable modules" in Ruby. I'd say that concerns are mixins so I feel that "In other languages, modules are often known as mixins" doesn't quite fit.

Reviewer 6: And that is the fundamental issue with documenting "concerns", concerns are modules, there's not a lot to say about them. You are a Ruby programmer, you know what modules are and their use cases. If you don't, you'll find in Ruby learning resources. Rails guides should not teach threads in Ruby, should not teach method dispatch in Ruby, and should not teach code reuse via mixins in Ruby.

Reviewer 6: In Ruby you split code in modules. If some module is reusable by several models, you put it in `app/models/concerns`. That's it.

On the other hand, Active Support has something called `ActiveSupport::Concern` that you do not need at all for writing concerns. A concern is a module. `AS::Concern` has some sugar for certain module inclusion patterns, and would have its own module documentation in the API.

And that is all.

Author: @fxn When I [first created this PR](https://github.com/rails/rails/pull/39766/commits/9aff8252f2e1dd65f52768c4da6719a7ffb9e480), given your previous feedback on the old PR, I actually had written the introduction as basically what you have here - concerns are modules, they help you reuse code. Does that seem like a better version than what is currently in the PR?

Aside from that, I think there's value in demonstrating how to use `ActiveSupport::Concern` or at least linking to the API docs from this guide. There's precedence for demonstrating the use of ActiveRecord or ActiveSupport in these guides - [here](https://guides.rubyonrails.org/active_record_basics.html) and [here](https://guides.rubyonrails.org/active_support_core_extensions.html) -- even though [they already have API guides](https://api.rubyonrails.org/classes/ActiveRecord/Base.html) - and from the feedback I'm seeing, there's value in having them as part of these guides as opposed to the API docs alone, which doesn't explicitly state the use for the folders. 

For me personally, what it comes down to is that a basic Rails app contains empty folders within prominent directories (`models` and `controllers`) that are not explicitly explained. If the folders were called `modules` instead of `concerns`, that would be pretty self-explanatory (even though I would personally still wonder why they are included if they are empty, when each user can create modules in any directory structure they like). Their presence and the name `concerns` implies a Rails convention. So as long as the folders exist with a name that is Rails-specific, we should have documentation somewhere intuitive to explain them - whether that's a tutorial like what I've written here, or just a paragraph or two (which the [previous PR was reduced to](https://github.com/rails/rails/pull/21114).) I totally agree with you that the Rails guides should not be responsible for demonstrating underlying Ruby concepts - but from what I can tell given the feedback I'm seeing, many people do not think `concerns` are just modules, and therefore they are not an underlying Ruby concept. 

I've always considered the Rails guides to be where you go to get wider context around how, when, and _why_ to use various Rails methods and conventions, whereas each page of the API guide is much more limited in scope. That is why I think the Rails guides should have some documentation for concerns (or remove the folders from Rails entirely, so we no longer imply a Rails convention if there isn't one). What the scope should be - I think that's the main challenge now. I can go either way - I think there is value in explaining the conceptual methodology, but if we can get a consensus on focusing these docs on modules without limiting the definition of concerns (as in @mcgregordan 's feedback) , I'm happy to go with that too. :) 

Tagging @sunny and @fylooi also - sorry this became a bit of an essay!

Reviewer 7: > I've always considered the Rails guides to be where you go to get wider context around how, when, and why to use various Rails methods and conventions

This is also exactly what I think of the Rails guides at and the proposed documentation changes in this PR already function very well for this purpose - it's easy to see how the examples given would extend to other domains and your own problems you're using rails to solve.

I just don't want to see the added value of what's there get overlooked in the chase for something that's perhaps even better, especially if that version involves explaining less than the current version does.

Reviewer 2: > On the other hand, Active Support has something called `ActiveSupport::Concern` that you do not need at all for writing concerns. A concern is a module. `AS::Concern` has some sugar for certain module inclusion patterns, and would have its own module documentation in the API.
> 
> And that is all.

While `ActiveSupport::Concern` can be used on its own, combining this sugar with the declarative DSL of an ActiveRecord model feels like a powerful and distinctly Rails convention to me. Think this particular pattern is worth highlighting.

Author: @kaspth @rafaelfranca what do you think of this PR so far? It would be great to hear if you share @fxn 's concerns around documentation or not - I made this PR based on feedback from others, but also your comments in my previous PR https://github.com/rails/rails/pull/21114 - so would be great to hear what you think! I haven't got any answer from a core team member since I responded.

> Concerns are a way to make large controllers or models easier to understand and manage. This also has the advantage of reusability when multiple models (or controllers) share the same concerns. Concerns are implemented using modules that contain methods representing a well-defined slice of the functionality that a model or controller is responsible for. In other languages, modules are often known as mixins.

Reviewer 8: I prefer the current wording.

>  You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog` , two folders were created within `app/` along with the rest:

Reviewer 5: ```suggestion
 You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog`, two folders were created within `app/` along with the rest:
```

> You can use concerns in your controller or model the same way you would use any module. When you first created your app with `rails new blog`, two folders were created within `app/` along with the rest:

Reviewer 8: I also prefer the current wording.

 ```
 app/controllers/concerns
 app/models/concerns
 ```

A given blog article might have various statuses - for instance, it might be visible to everyone (i.e. `public`), or only visible to the author (i.e. `private`). It may also be hidden to all but still retrievable (i.e. `archived`). Comments may similarly be hidden or visible. This could be represented using a `status` column in each model.

Within the `article` model, after running a migration to add a `status` column, you might add:

```ruby
class Article < ApplicationRecord
  has_many :comments
  validates :title, presence: true,

>                     length: { minimum: 5 }

Author: @mcgregordan I see what you're saying but I don't think I should change this, because this is copied from the earlier model code snippets in the Getting Started guide. Someone following the guide would get confused if we're talking about the same Article and Comment models that we were discussing in the previous sections of the guide, but then suddenly they were missing their validations and such.

  validates :status, in: VALID_STATUSES

  def archived? 
    status == 'archived'
  end
end
```

and in the `Comment` model:

```ruby
class Comment < ApplicationRecord
  belongs_to :article

  VALID_STATUSES = ['public', 'private', 'archived']

  validates :status, in: VALID_STATUSES

  def archived?
    status == 'archived'
  end
end
```

Then, in our `index` action template (`app/views/articles/index.html.erb`) we would use the `archived?` method to avoid displaying any article that is archived:

```html+erb
<h1>Listing Articles</h1>
<%= link_to 'New article', new_article_path %>
<table>
  <tr>
    <th>Title</th>
    <th>Text</th>
    <th colspan="3"></th>
  </tr>

  <% @articles.each do |article| %>
    <% unless article.archived? %> 
      <tr>
        <td><%= article.title %></td>
        <td><%= article.text %></td>
        <td><%= link_to 'Show', article_path(article) %></td>
        <td><%= link_to 'Edit', edit_article_path(article) %></td>
        <td><%= link_to 'Destroy', article_path(article),
                method: :delete,
                data: { confirm: 'Are you sure?' } %></td>
      </tr>
    <% end %>
  <% end %>
</table>
```

> However, if you look again at our models now, you can see that the logic is duplicated. If in future we increase the functionality of our blog - to include private messages, for instance -  we might find ourselves duplicating the logic yet again. This is where concerns come in handy.

Reviewer 5: ```suggestion
However, if you look again at our models now, you can see that the logic is duplicated. If in the future we increase the functionality of our blog - to include private messages, for instance -  we might find ourselves duplicating the logic yet again. This is where concerns come in handy.
```

> A concern is only responsible for a focused subset of the model's responsibility; the methods in our concern will all be related to the visibility of a model. Let's call our new concern (module) `Visible`. We can create a new file inside `app/models/concerns` called `visible.rb` , and store all of the status methods that were duplicated in the models.

Author: @fylooi I amended this paragraph so that it mentioned that a concern has to have a focused responsibility - I realized as I made the change you requested earlier up that my paragraph here needed a bit more work so people understand that you don't want to just dump any model methods in a single concern. :)

`app/models/concerns/visible.rb`

```ruby
module Visible
  def archived?
    status == 'archived'
  end
end
```

We can add our status validation to the concern, but this is slightly more complex as validations are methods called at the class level. The `ActiveSupport::Concern` ([API Guide](https://api.rubyonrails.org/classes/ActiveSupport/Concern.html)) gives us a simpler way to include them:

```ruby
module Visible
  extends ActiveSupport::Concern

  included do
    VALID_STATUSES = ['public', 'private', 'archived']

    validates :status, in: VALID_STATUSES
  end

  def archived?
    status == 'archived'
  end

> end

Reviewer 7: I think using the same name for different parts of an example adds an unnecessary overhead in understanding what's going on. Someone new might look at a concern named `Status` that operates on an attribute named `status` and assume a stronger link between these two things than actually exists. Having the concern be named something different makes it much clearer how the two things relate.

> end

Author: I agree with @tvanderpol on this one :) I actually picked `Visible` as I noticed it was one of the example concern names in a separate blogpost about concerns and I thought it made sense for this functionality.

```

Now, we can remove the duplicated logic from each model and instead include our new `Visible` module:

In `app/models/article.rb`:
```ruby
class Article < ApplicationRecord
  include Visible
  has_many :comments

  validates :title, presence: true,
                    length: { minimum: 5 }

end
```

and in `app/models/comment.rb`:

```ruby
class Comment < ApplicationRecord
  include Visible
  belongs_to :article
end
```

> Class methods can also added to concerns. If we want a count of public articles or comments to display on our main page, we might add a class method to Visible as follows:

Author: @dzunk I added this section regarding class methods - I admit to struggling with an example that works in the context of this Getting Started blog. I would have preferred to make a class method that retrieves all public articles/comments by user, but there is no `user` object in this blog app - the closest we get to that is having a `commenter`.  Would be great to hear your thoughts on this!

Reviewer: I think it's a fine example, although you might want to update the logic to actually `.count` the records 😉 

An alternative could be something related to the state transitions:
```ruby
class_methods do
  def archive_all
    update_all(status: 'archived')
  end
end
```

Author: Ahh, good catch! 😅  I quite like your alternative method, but it seems to me like the usage example would be a little more complicated (one would probably use the method in a new action) so I'll keep the `count` example (and make sure it actually does count... 🙈 )

> Class methods can also added to concerns. If we want a count of public articles or comments to display on our main page, we might add a class method to Visible as follows:

Reviewer 5: ```suggestion
Class methods can also be added to concerns. If we want a count of public articles or comments to display on our main page, we might add a class method to Visible as follows:
```

```ruby
module Visible
  extends ActiveSupport::Concern

  VALID_STATUSES = ['public', 'private', 'archived']

  included do
    validates :status, in: VALID_STATUSES
  end

  class_methods do
    def public_count

>       self.where(status: 'public')

Reviewer 7: This probably needs a `.count` added.

Author: I totally missed it 🙈 This should be fixed now!

>       self.where(status: 'public').count

Reviewer 5: We can drop the superflous `self.` here.

```suggestion
      where(status: 'public').count
```

```ruby
class Article < ApplicationRecord

>   includes Statusable

Reviewer 4: ```suggestion
  include Statusable
```
